src/components/ChildNodeViews.tsx

```diff
@@ -0,0 +1,364 @@
+  cloneElement,
+  createElement,
+  useContext,
+} from "react";
+
+import { ChildDescriptorsContext } from "../contexts/ChildDescriptorsContext.js";
+import { EditorContext } from "../contexts/EditorContext.js";
+import {
+  NonWidgetType,
+  ReactWidgetDecoration,
+} from "../decorations/ReactWidgetType.js";
+import { iterDeco } from "../decorations/iterDeco.js";
+import { useEditorState } from "../hooks/useEditorState.js";
+import { useReactKeys } from "../hooks/useReactKeys.js";
+import {
+  Decoration,
+  DecorationSource,
+} from "../prosemirror-view/decoration.js";
+
+import { MarkView } from "./MarkView.js";
+import { NodeView } from "./NodeView.js";
+import { TextNodeView } from "./TextNodeView.js";
+import { TrailingHackView } from "./TrailingHackView.js";
+import { WidgetView } from "./WidgetView.js";
+
+function cssToStyles(css: string) {
+  const cssJson = `{"${css
+    .replace(/;? *$/, "")
+    .replace(/;+ */g, '","')
+    .replace(/: */g, '":"')}"}`;
+
+  const obj = JSON.parse(cssJson);
+
+  return Object.keys(obj).reduce((acc, key) => {
+    const camelCased = key.startsWith("--")
+      ? key
+      : key.replace(/-[a-z]/g, (g) => g[1]?.toUpperCase() ?? "");
+    return { ...acc, [camelCased]: obj[key] };
+  }, {});
+}
+
+export function wrapInDeco(reactNode: JSX.Element | string, deco: Decoration) {
+  const {
+    nodeName,
+    class: className,
+    style: css,
+    contenteditable: contentEditable,
+    spellcheck: spellCheck,
+    ...attrs
+  } = (deco.type as unknown as NonWidgetType).attrs;
+
+  // We auto-wrap text nodes in spans so that we can apply attributes
+  // and styles, but we want to avoid double-wrapping the same
+  // text node
+  if (nodeName || typeof reactNode === "string") {
+    return createElement(
+      nodeName ?? "span",
+      {
+        className,
+        contentEditable,
+        spellCheck,
+        style: css && cssToStyles(css),
+        ...attrs,
+      },
+      reactNode
+    );
+  }
+
+  return cloneElement(reactNode, {
+    className: classnames(reactNode.props.className, className),
+    contentEditable,
+    spellCheck,
+    style: { ...reactNode.props.style, ...(css && cssToStyles(css)) },
+    ...attrs,
+  });
+}
+
+type ChildWidget = {
+  type: "widget";
+  widget: ReactWidgetDecoration;
+  marks: readonly Mark[];
+  offset: number;
+  index: number;
+};
+
+type ChildNode = {
+  type: "node";
+  node: Node;
+  marks: readonly Mark[];
+  innerDeco: DecorationSource;
+  outerDeco: readonly Decoration[];
+  offset: number;
+};
+
+type ChildTrailingHack = {
+  type: "trailinghack";
+  offset: 0;
+};
+
+type Child = ChildNode | ChildWidget;
+
+type SharedMarksProps = {
+  innerPos: number;
+  childViews: Child[];
+};
+
+function InlineView({ innerPos, childViews }: SharedMarksProps) {
+  const { editorView } = useContext(EditorContext);
+  const editorState = useEditorState();
+  const reactKeys = useReactKeys();
+
+  const partitioned = childViews.reduce((acc, child) => {
```

Comment:
trying to wrap my head around this...this attempts to group like mark nodes and text widgets by mark type?

Reply:
Yeah this was gnarly and is probably not the most legible way to express it. There are some really complex rules that ProseMirror uses to flatten/simplify/group marks, and I wanted to be sure that we reproduced exactly the same behavior.

Here's what I wrote in the Gitter channel when I was working on this:

> A thing that I learned: Though inline model nodes are represented as flat nodes with marks, when possible, adjacent nodes with the same outer decorations and marks are actually merged in the HTML output. So for example:
>
>```
>doc
>  p
>    "bold and italic" #text [em, strong]
>    "italic" #text [em]
>```
>
> Is rendered as:
>
>```html
><div>
>  <p>
>    <em>
>      <strong>bold and italic</strong>
>      italic
>    </em>
>  </p>
></div>
>```

Reply:
Ok well they're not THAT complex, they're just non-trivial to express with React haha